sonyxperiadev/kernel
# [6.4.r1] Add Clang support [1/2]

Part 2 is in [qcacld-3.0](https://github.com/sonyxperiadev/vendor-qcom-opensource-wlan-qcacld-3.0/pulls).

This supersedes #1625.

* **Why even do this?**

  * Google has been moving away from GCC for a few years now (AOSP userland is completely compiled with Clang). They currently compile the Pixel 2 (XL) kernel with Clang 5.0 and will compile all 4.4 Chromebooks with Clang 7.0 on R67 (currently in dev, soon in beta - [source 1](https://chromium-review.googlesource.com/809774), [source 2](https://lkml.org/lkml/2018/4/3/567)).
  * Clang offers some nice benefits like addition warnings, better static analysis, and hopefully in the near future, nice optimizations to improve code execution.
  * The ability to do side by side builds with `-Werror` ([like so](https://github.com/nathanchance/scripts/commit/2b3f8251578d564b559d93b01013159e42ebd4b8)) to ensure the highest code quality.
  * Google gave a talk about their journey to do this ([a video](https://www.youtube.com/watch?v=6l4DtR5exwo), [an article](https://lwn.net/Articles/734071/)), highly recommended!

* **Technical details**

  * The first section, the core patchset, comes from mainline, with the backport being done by Chromium ([git repo](https://chromium.googlesource.com/chromiumos/third_party/kernel/+log/sandbox/mka/llvm/v4.4)).
  * The second section comes from the Pixel 2 kernel, focusing on fixing out of tree code to properly compile with Clang.
  * The third section comes from mainline, focusing on fixing warnings that come from Clang for in tree code.
  * The fourth section comes from me, focusing on fixing warnings that come from Clang for out of tree code. Google hid a lot of warnings that were very simple to fix, I went ahead and did it.
  * Commit suffixes:
    * `UPSTREAM`: A commit that comes from mainline and was a clean pick.
    * `BACKPORT`: A commit that comes from mainline and had conflicts.
    * `FROMLIST`: A commit that was sent to a mailing list (may or may not have been accepted upstream).
    * `HACK`: A solution for a problem that may or may not be ideal and probably deserves some scrutiny.

* **How to test**

  1. Download the master branch of [Clang](https://android.googlesource.com/platform/prebuilts/clang/host/linux-x86/) and [GCC](https://android.googlesource.com/platform/prebuilts/gcc/linux-x86/aarch64/aarch64-linux-android-4.9/) (newer binutils) from AOSP.
  2. Run the following commands in your kernel source:
  ```bash
  make ARCH=arm64 O=out <config_name>
  make -j$(nproc --all) \
       ARCH=arm64 \
       CC="<path_to_clang_folder>/bin/clang" \
       CLANG_TRIPLE=aarch64-linux-gnu- \
       CROSS_COMPILE="<path_to_gcc_folder>/bin/aarch64-linux-android-" \
       O=out
  ```
  3. If you want to make sure the build successfully compiled with Clang before flashing, open the `out/include/generated/compile.h` file and look at the `LINUX_COMPILER` option.
    * NOTE: On Google's versions of Clang, there are URLs to the HEAD commit of both LLVM and Clang). To get rid of these, a follow-up commit is needed to allow the version string to be customized. If that is of interest, I will add it to this patchset.
    
This patchset has been build tested against all of the `aosp_*` defconfigs for GCC 4.9.4, GCC 7.3.0, Clang 6.0 (clang-4691093), and Clang 7.0 (clang-4679922) with `CONFIG_CC_WERROR` turned on.
